File: backend/src/routes/tracking.js
import { Router } from 'express';
import prisma from '../lib/prisma.js';

const router = Router();

// GET /api/track/:pixelId — serves a 1x1 transparent GIF and records the open
// The tracking pixel URL is embedded in outgoing emails as an <img> tag.
const TRANSPARENT_GIF = Buffer.from(
  'R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7',
  'base64'
);

router.get('/:pixelId', async (req, res) => {
  try {
    // Fire-and-forget: update the email's openedAt timestamp
    await prisma.email.updateMany({
      where: { trackingPixelId: req.params.pixelId, openedAt: null },
      data: { openedAt: new Date(), status: 'opened' },
    });
  } catch (err) {
    console.error('Tracking pixel error:', err);
  }

  // Always return the image regardless of DB errors
  res.set({
    'Content-Type': 'image/gif',
    'Content-Length': TRANSPARENT_GIF.length,
    'Cache-Control': 'no-store, no-cache, must-revalidate, proxy-revalidate',
  });
  res.end(TRANSPARENT_GIF);
});

export default router;

File: app/routes/tracking.py
"""
Tracking pixel route.
GET /track/:pixel_id — serves a 1x1 transparent GIF and records the open
"""

import base64
from datetime import datetime
from fastapi import APIRouter, Depends
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.database import get_db
from app.models import Email

router = APIRouter()

# 1x1 transparent GIF (43 bytes)
PIXEL = base64.b64decode(
    "R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
)


@router.get("/{pixel_id}")
async def track_open(pixel_id: str, db: AsyncSession = Depends(get_db)):
    """Record an email open event when the tracking pixel is loaded."""
    email = (await db.execute(
        select(Email).where(Email.tracking_pixel_id == pixel_id)
    )).scalar_one_or_none()

    if email and email.status in ("sent", "follow_up_sent"):
        email.status = "opened"
        email.opened_at = datetime.utcnow()
        await db.flush()

    return Response(content=PIXEL, media_type="image/gif")
